crates/pathfinder/src/state.rs

```diff
@@ -40,167 +42,415 @@ enum UpdateError {
+        let definition = compressor
+            .compress(&*definition)
+            .context("Failed to compress definition")?;
+
+        Result::<_, anyhow::Error>::Ok(CompressedContract {
+            abi,
+            bytecode,
+            definition,
+            hash,
+        })
+    };
+
+    while let Some((deployed_contract, def_bytes)) = rx.blocking_recv() {
+        tx.blocking_send((deployed_contract, process_one(def_bytes)))
+            .unwrap();
+    }
+}
+
+struct CompressedContract {
+    abi: Vec<u8>,
+    bytecode: Vec<u8>,
+    definition: Vec<u8>,
+    hash: StarkHash,
+}
+
+impl std::fmt::Debug for CompressedContract {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "CompressedContract {{ sizes: {:?}, hash: {} }}",
+            (self.abi.len(), self.bytecode.len(), self.definition.len()),
+            self.hash
+        )
     }
 }
 
 /// Updates the Starknet state with a new block described by [StateUpdateLog].
 ///
 /// Returns the new global root.
-async fn update<T: Transport>(
-    transport: &Web3<T>,
+fn update(
+    state_update: StateUpdate,
     global_root: GlobalRoot,
     update_log: &StateUpdateLog,
     db: &Transaction<'_>,
-    sequencer: &sequencer::Client,
-) -> Result<GlobalStateRecord, UpdateError> {
+    fetched_contracts_tx: &mpsc::Sender<Vec<DeployedContract>>,
+    ready_contracts_rx: &mut mpsc::Receiver<(DeployedContract, anyhow::Result<CompressedContract>)>,
+) -> Result<BlockUpdated, UpdateError> {
     // Download update from L1.
-    use RetrieveStateUpdateError::*;
-    let state_update = match StateUpdate::retrieve(transport, update_log.clone()).await {
-        Ok(state_update) => state_update,
-        Err(Other(other)) => {
-            return Err(UpdateError::Other(anyhow::anyhow!(
-                "Fetching state update failed. {}",
-                other
-            )));
-        }
-        // Treat the rest as a reorg event.
-        Err(_reorg) => return Err(UpdateError::Reorg),
-    };
+
+    let contract_count = state_update.deployed_contracts.len();
+
+    // TODO: check if the contract has been downloaded already
+    fetched_contracts_tx
+        .blocking_send(state_update.deployed_contracts)
+        .unwrap();
 
     let mut global_tree =
         GlobalStateTree::load(db, global_root).context("Loading global state tree")?;
 
-    // Deploy contracts
-    for contract in state_update.deployed_contracts {
-        // Add new contract's to global tree, the contract roots are initialized to ZERO.
-        let contract_root = ContractRoot(StarkHash::ZERO);
-        let state_hash = calculate_contract_state_hash(contract.hash, contract_root);
-        global_tree
-            .set(contract.address, state_hash)
-            .context("Adding deployed contract to global state tree")?;
-        ContractsStateTable::insert(db, state_hash, contract.hash, contract_root)
+    for _ in 0..contract_count {
```

Comment:
I think this didn't need to happen right away, the contract updates could be done first, and this could be only before the root is applied. It'd save writing those zero contractstates, and give more time for the compression ... though we have only queue of size one.

Reply:
You would just need to track which contracts have updates and which don't. Guess one could hashmap them with default ZERO or something.